### DISCUSSION TITLE
Better TEST_OUTPUT

### DISCUSSION THREAD

**Discussion by dkorpel** [dlang/dmd#14515] [General]
When you have a big fail_compilation test, for example:

https://github.com/dlang/dmd/blob/master/compiler/test/fail_compilation/retscope6.d

There are two annoyances:

- Because of multiple `TEST_OUTPUT` blocks, they cannot be automatically updated with `AUTO_UPDATE=1` 
- Because of `#line` directives for stable lines, you cannot quickly locate an error's corresponding line in your text editor.

(See also https://github.com/dlang/dmd/pull/14434#discussion_r973763212 and https://github.com/dlang/dmd/pull/14434#discussion_r973762966 where this came up)

Splitting it up into smaller tests isn't all so great either:
- The huge amount of test files is annoying to browse. The GitHub web viewer even truncates it to the first 1000.
- Every file gives a bit of overhead making the test suite run slower

But what if instead of `TEST_OUTPUT` blocks, the test runner could also parse `// Error:` or `// Deprecation:` comments:

```D
int* test() @safe
{
    int i;
    int*[][] arr = new int*[][](1);
// Error: copying `& i` into allocated memory escapes a reference to local variable `i`
    arr[0] ~= &i;
    return arr[0][0];
}
```

The whole `fail_compilation/retscope6.d(6007):` prefix will be inferred from the location of the comment in the source file. I think this is both easy to write, read, and mechanically auto update. What do you think?

**Comment by Geod24** [comment]
I think it should be marked with some kind of unmistakable marker (e.g. `@@TEST@@`) so that you can still copy-paste old error message without having false positive, but it's not a big point either. I'm all for making testing better and it looks like this would.

**Reply by ibuclaw** [reply]
For comparison, this is an example of marking errors in dejagnu (also, if an error occurs twice on the same line).
https://github.com/D-Programming-GDC/gcc/blob/ci/mainline/gcc/testsuite/gdc.dg/asm1.d#L8-L19

**Reply by maxhaton** [reply]
https://github.com/ldc-developers/ldc/blob/master/tests/fail_compilation/betterC_typeinfo_diag.d#L8 ldc/llvm equiv

**Comment by ljmf00-wekaio** [comment]
Why don't we use `lit` with `FileCheck`? LLVM does this for years and it's much more intuitive. See https://llvm.org/docs/CommandGuide/FileCheck.html

One example:
```
; RUN: llvm-as < %s | llc -mtriple=i686-apple-darwin9 -mattr=sse41 \
; RUN:              | FileCheck %s -check-prefix=X32
; RUN: llvm-as < %s | llc -mtriple=x86_64-apple-darwin9 -mattr=sse41 \
; RUN:              | FileCheck %s -check-prefix=X64

define <4 x i32> @pinsrd_1(i32 %s, <4 x i32> %tmp) nounwind {
        %tmp1 = insertelement <4 x i32>; %tmp, i32 %s, i32 1
        ret <4 x i32> %tmp1
; X32: pinsrd_1:
; X32:    pinsrd $1, 4(%esp), %xmm0

; X64: pinsrd_1:
; X64:    pinsrd $1, %edi, %xmm0
}
```

**Reply by ljmf00-wekaio** [reply]
One of the geat things that FileCheck adds is the ability to only partial match the output, so we don't need to specify the whole output there. This makes it easier to update error messages, when the compiler gives errors unrelated to the test.

**Reply by ljmf00-wekaio** [reply]
FileCheck works for everything that has an output, its not tied to any specific output standard. Its pure pattern matching checks. So we can add tests to the output of `-vasm` e.g. or other things, like dwarf backend implementation.
I see dwarf tests particularly janky, when I did changes there.

**Reply by ljmf00-wekaio** [reply]
CC @JohanEngelen @kinke maybe you can help me supporting this idea with suggestions over our current testsuite in DMD :)

**Reply by JohanEngelen** [reply]
`lit` and `FileCheck` and `not` (and `split-file`) that LDC uses are *way* better than the current dmd testsuite. Fully support converting to that.

**Reply by dkorpel** [reply]
> much more intuitive

Idk, that example you gave pretty much looks like gibberish to me :laughing: 

But I do like the idea of using a common tool instead of an own invention, so it's worth a look at.

**Reply by ljmf00** [reply]
> Idk, that example you gave pretty much looks like gibberish to me 😆

Probably because its matching assembly but its pretty intuitive if you look at the link I sent.
